api/app/playwright_renderer.py

```diff
@@ -0,0 +1,106 @@
+import uuid
+from typing import Final, Optional
+from urllib.parse import parse_qs, urlparse
+
+from playwright.async_api import async_playwright
+
+# Html selectors
+LAYER_ACCORDION_SELECTOR: Final[
+    str
+] = 'div[class="MuiAccordionSummary-root"], div[aria-expanded="false"]'
+DOWNLOAD_BUTTON_SELECTOR: Final[str] = "a[download]"
+CREATE_REPORT_BUTTON_SELECTOR: Final[str] = 'button[id="create-report"]'
+
+# Timeouts
+PAGE_TIMEOUT: Final[int] = 25000
+PAGE_LANGUAGE_CHANGE_TIMEOUT: Final[int] = 5000
+
+
+async def playwright_download_report(url: str, language: Optional[str]) -> str:
+    async with async_playwright() as p:
+        browser = await p.chromium.launch()
+        page = await browser.new_page()
+
+        page.set_default_timeout(PAGE_TIMEOUT)
+        await page.goto(url)
+
+        layerIdParam = extract_layerId_query_param(url)
+
+        # Wait for page to be loaded
+        await page.wait_for_selector(LAYER_ACCORDION_SELECTOR, state="visible")
+
+        # Toggle level one dropdowns
+        await toggle_every_visible_dropdown(page)
+
+        # Toggle level two dropdowns
+        await toggle_every_visible_dropdown(page)
+
+        # Click on exposure analysis toggle button
+        await click_target_exposure_analysis(page, layerIdParam)
+
+        # Wait for page to be loaded on exposure analysis
+        await page.wait_for_selector(CREATE_REPORT_BUTTON_SELECTOR, state="visible")
+
+        # Change language if not english
+        await change_language_if_not_default(language, page)
+
+        await click_create_report_button(page)
+
+        # Wait for report to be created by pdf-renderer
+        await page.wait_for_selector(DOWNLOAD_BUTTON_SELECTOR)
+
+        async with page.expect_download() as download_info:
+            download_report_selector = await page.query_selector(
+                DOWNLOAD_BUTTON_SELECTOR
+            )
+            await download_report_selector.click()
+
+        download = await download_info.value
+        random_uuid = uuid.uuid4()
+        report_file_path = (
+            "./report-"
+            + layerIdParam
+            + "-"
+            + language
+            + "-"
+            + str(random_uuid)
+            + ".pdf"
+        )
+
+        await download.save_as(report_file_path)
```

Comment:
You should know reports are going to pile up in the production container.
- I didn't implemented a cleanup since each new API deployment should remove those files
- In the long run we might implement a cache strategy for them

Reply:
I don't think new API deployments actually remove any files. But it's a reasonable idea

Reply:
You could also type your functions when possible

Reply:
I made my comment before I implemented the logic that put reports in the `/cache` volume. So yes now they won't be removed